File: src/lib/tags.ts
// Hashtag detection + the canonical-tag-line relocation transform.
// Mirrors src-tauri/src/tags.rs — keep the two in sync.
//
// Grammar:
//   - #tag must follow whitespace, start-of-line, or one of [({"'.
//   - Starts with an ASCII letter, then [a-z0-9_/-].
//   - Lowercase canonical.
//   - URL fragments (https://...#x) and code blocks/spans are skipped.

export type TagMatch = {
  from: number; // char offset in the original string
  to: number; // exclusive
  tag: string; // canonical (lowercase)
};

const OPEN_LEAD = new Set(["(", "[", "{", '"', "'"]);

function isLetter(c: string): boolean {
  return /^[a-z]$/i.test(c);
}
function isTagChar(c: string): boolean {
  return /^[a-z0-9_/-]$/i.test(c);
}
function isValidLead(prev: string | undefined): boolean {
  if (prev === undefined) return true;
  if (/\s/.test(prev)) return true;
  return OPEN_LEAD.has(prev);
}

/**
 * Find every inline hashtag in `body`, respecting code fences + inline code.
 * Returns matches in body order (deduplicated by position, not by tag value).
 */
export function findInlineTags(body: string): TagMatch[] {
  const out: TagMatch[] = [];
  let lineStart = 0;
  let inFence = false;
  for (let i = 0; i <= body.length; i++) {
    if (i === body.length || body[i] === "\n") {
      const line = body.slice(lineStart, i);
      const trimmed = line.trimStart();
      if (trimmed.startsWith("```") || trimmed.startsWith("~~~")) {
        inFence = !inFence;
      } else if (!inFence) {
        scanLineInto(line, lineStart, out);
      }
      lineStart = i + 1;
    }
  }
  return out;
}

function scanLineInto(line: string, lineOffset: number, out: TagMatch[]): void {
  // Mask inline-code regions with spaces so offsets stay aligned.
  const masked = maskInlineCode(line);
  for (let i = 0; i < masked.length; i++) {
    if (masked[i] !== "#") continue;
    if (!isValidLead(i === 0 ? undefined : masked[i - 1])) continue;
    const start = i + 1;
    if (start >= masked.length) break;
    if (!isLetter(masked[start])) continue;
    let j = start + 1;
    while (j < masked.length && isTagChar(masked[j])) j++;
    // Strip trailing slashes — partial nesting isn't a tag.
    let end = j;
    while (end > start && masked[end - 1] === "/") end--;
    if (end > start) {
      const tag = masked.slice(start, end).toLowerCase();
      out.push({ from: lineOffset + i, to: lineOffset + end, tag });
    }
    i = j - 1;
  }
}

function maskInlineCode(line: string): string {
  let inCode = false;
  let out = "";
  for (const c of line) {
    if (c === "`") {
      inCode = !inCode;
      out += " ";
    } else if (inCode) {
      out += " ";
    } else {
      out += c;
    }
  }
  return out;
}

/**
 * Returns true if `line` is entirely hashtags + whitespace.
 */
export function isCanonicalTagLine(line: string): boolean {
  const trimmed = line.trim();
  if (!trimmed) return false;
  const toks = trimmed.split(/\s+/);
  for (const tok of toks) {
    if (!tok.startsWith("#")) return false;
    const body = tok.slice(1);
    if (!body || !isLetter(body[0])) return false;
    for (const c of body.slice(1)) {
      if (!isTagChar(c)) return false;
    }
  }
  return true;
}

/**
 * Locate the canonical tag line if present. Returns the line index (0-based)
 * and the parsed tag set, or null. The "canonical" line is the last
 * non-empty line in the body that's entirely hashtags.
 */
export function findCanonicalTagLine(body: string): { lineIdx: number; tags: string[] } | null {
  const lines = body.split("\n");
  for (let i = lines.length - 1; i >= 0; i--) {
    const l = lines[i];
    if (l.trim() === "") continue;
    if (isCanonicalTagLine(l)) {
      const tags = l
        .trim()
        .split(/\s+/)
        .map((t) => t.replace(/^#/, "").toLowerCase());
      return { lineIdx: i, tags };
    }
    return null; // first non-empty line from bottom is something else
  }
  return null;
}

/**
 * Given the current body, produce a normalized version where:
 *   - All inline #tags above the canonical line are removed (with adjacent
 *     whitespace collapsed sensibly)
 *   - All discovered tags are merged with the existing canonical-line tags
 *   - The canonical tag line is rewritten as the last non-empty line
 *
 * Returns the new body + the unified tag set. If no changes are needed,
 * returns the input body unchanged so the editor can short-circuit.
 */
export function relocateTagsToBottom(body: string): { body: string; tags: string[] } {
  const canonical = findCanonicalTagLine(body);
  const canonicalIdx = canonical?.lineIdx ?? -1;
  const lines = body.split("\n");

  const aboveText = canonicalIdx === -1 ? body : lines.slice(0, canonicalIdx).join("\n");
  const inline = findInlineTags(aboveText);

  // If there are no inline tags above and the canonical line is already at
  // the bottom (no trailing blank lines past it), nothing to do.
  if (inline.length === 0) {
    if (canonicalIdx === -1) return { body, tags: [] };
    // Check that nothing nonempty comes after the canonical line.
    const tail = lines.slice(canonicalIdx + 1).join("\n").trim();
    if (tail === "") {
      return { body, tags: canonical!.tags };
    }
  }

  // Rewrite the above-text with inline tags removed.
  let newAbove = "";
  let cursor = 0;
  for (const m of inline) {
    newAbove += aboveText.slice(cursor, m.from);
    // Collapse one adjacent space — prefer the trailing space if the
    // following char is whitespace; otherwise consume one leading space.
    const after = aboveText[m.to];
    if (after === " " || after === "\t") {
      cursor = m.to + 1;
    } else if (
      newAbove.length > 0 &&
      (newAbove.endsWith(" ") || newAbove.endsWith("\t"))
    ) {
      newAbove = newAbove.slice(0, -1);
      cursor = m.to;
    } else {
      cursor = m.to;
    }
  }
  newAbove += aboveText.slice(cursor);
  // Trim trailing whitespace on every line, collapse runs of >2 blank lines.
  newAbove = newAbove
    .split("\n")
    .map((l) => l.replace(/[ \t]+$/, ""))
    .join("\n")
    .replace(/\n{3,}/g, "\n\n");

  // Merge tags: from existing canonical line + freshly extracted.
  const allTags = new Set<string>();
  for (const t of canonical?.tags ?? []) allTags.add(t);
  for (const m of inline) allTags.add(m.tag);
  const sorted = [...allTags].sort();

  // Compose the new body. If there are tags, emit a canonical tag line at
  // the very bottom, separated by a blank line if the body has content.
  let result = newAbove.replace(/\n+$/, "");
  if (sorted.length > 0) {
    if (result.length > 0) result += "\n\n";
    result += sorted.map((t) => `#${t}`).join(" ");
  }
  result += "\n";

  return { body: result, tags: sorted };
}

/**
 * Strip wikilink brackets, replacing `[[X]]` with bare `X`. The AI shouldn't
 * see our private markup; the inner text IS the prose. Multi-line wikilinks
 * aren't a thing (the resolver doesn't allow `\n` inside the brackets), so
 * a same-line regex is sufficient. Nested `[` is also disallowed.
 */
export function stripWikilinkBrackets(text: string): string {
  return text.replace(/\[\[([^\[\]\n]+)\]\]/g, "$1");
}

/**
 * Strip the canonical tag line + every inline hashtag + every `[[...]]`
 * bracket from `text`. Used before sending content to the AI so our private
 * markup doesn't pollute the model context. Whitespace around removed tags
 * is collapsed sensibly; blank-line runs are clamped to a single blank.
 */
export function stripTagsForAI(text: string): string {
  // 0) Strip wikilink brackets first so subsequent passes operate on pure
  //    prose. `[[Atomic Habits]]` → `Atomic Habits`.
  text = stripWikilinkBrackets(text);
  // 1) Drop the canonical tag line if it ends `text` (last non-empty line).
  const lines = text.split("\n");
  let lastNonempty = -1;
  for (let i = lines.length - 1; i >= 0; i--) {
    if (lines[i].trim() !== "") {
      lastNonempty = i;
      break;
    }
  }
  if (lastNonempty >= 0 && isCanonicalTagLine(lines[lastNonempty])) {
    lines.splice(lastNonempty, 1);
    // Also drop a single preceding blank to avoid trailing blank lines.
    if (lastNonempty - 1 >= 0 && lines[lastNonempty - 1].trim() === "") {
      lines.splice(lastNonempty - 1, 1);
    }
  }
  let result = lines.join("\n");

  // 2) Remove every remaining inline hashtag, collapsing adjacent whitespace.
  const inline = findInlineTags(result);
  let out = "";
  let cursor = 0;
  for (const m of inline) {
    out += result.slice(cursor, m.from);
    const after = result[m.to];
    if (after === " " || after === "\t") {
      cursor = m.to + 1;
    } else if (out.length > 0 && (out.endsWith(" ") || out.endsWith("\t"))) {
      out = out.slice(0, -1);
      cursor = m.to;
    } else {
      cursor = m.to;
    }
  }
  out += result.slice(cursor);

  return out
    .split("\n")
    .map((l) => l.replace(/[ \t]+$/, ""))
    .join("\n")
    .replace(/\n{3,}/g, "\n\n");
}

/**
 * Find the range of the canonical tag line in the body (start and end char
 * offsets in the full document). Returns null if there is no canonical
 * tag line. Used by editor decorations to hide the line.
 */
export function canonicalTagLineRange(body: string): { from: number; to: number } | null {
  const lines = body.split("\n");
  let offset = 0;
  let lastNonempty = -1;
  let lastNonemptyOffset = 0;
  let lastNonemptyLen = 0;
  for (let i = 0; i < lines.length; i++) {
    const l = lines[i];
    if (l.trim() !== "") {
      lastNonempty = i;
      lastNonemptyOffset = offset;
      lastNonemptyLen = l.length;
    }
    offset += l.length + 1; // +1 for the newline
  }
  if (lastNonempty < 0) return null;
  if (!isCanonicalTagLine(lines[lastNonempty])) return null;
  // Refuse to hide the canonical line if there is no non-empty content
  // above it — otherwise the user sees nothing and the cursor has no
  // valid landing spot. This shows up on brand-new notes whose only
  // content so far is a hashtag, or on legacy notes that were just tags.
  let hasContentAbove = false;
  for (let i = 0; i < lastNonempty; i++) {
    if (lines[i].trim() !== "") {
      hasContentAbove = true;
      break;
    }
  }
  if (!hasContentAbove) return null;
  return { from: lastNonemptyOffset, to: lastNonemptyOffset + lastNonemptyLen };
}

File: src-tauri/src/tags.rs
// Hashtag extraction + the canonical "tag line" convention.
//
// Storage model:
//   - The canonical tag line is the last non-empty line of the body when it
//     matches `^\s*(#tag(\s+#tag)*)\s*$`. The editor moves any inline #tags
//     into this line on autosave so the body stays clean of tag-clutter.
//   - We still recognize inline #tags anywhere (so a freshly typed tag is
//     found before relocation runs). YAML frontmatter `tags:` is read for
//     backward compat with Obsidian / Bear / nvalt files but never written.
//
// Tag grammar:
//   - Must follow whitespace, start-of-line, or one of [({\[
//   - Starts with `#` then an ASCII letter, then [a-z0-9_/-]*
//   - Bear-style nesting via `/`: #projects/malt — query `tag:projects` ALSO
//     matches `#projects/malt` (we index every slash-prefix).
//   - URLs and code blocks/spans are skipped.

use crate::frontmatter;

/// Extract a deduplicated, lowercase, sorted list of tags from a note body.
/// Includes both inline #tags and the canonical tag line; YAML frontmatter
/// `tags:` is folded in by the caller (which has access to the parsed
/// frontmatter) — this fn deals with the body only.
pub fn extract_tags_from_body(body: &str) -> Vec<String> {
    let mut out = Vec::new();
    let mut in_fence = false;
    for line in body.lines() {
        let trimmed = line.trim_start();
        if trimmed.starts_with("```") || trimmed.starts_with("~~~") {
            in_fence = !in_fence;
            continue;
        }
        if in_fence {
            continue;
        }
        let cleaned = strip_inline_code(line);
        for tag in scan_line(&cleaned) {
            out.push(tag);
        }
    }
    out.sort();
    out.dedup();
    out
}

/// Like `extract_tags_from_body`, but also folds in YAML frontmatter tags.
/// Use this when you want the full set of tags for indexing / display.
pub fn extract_tags_full(content: &str) -> Vec<String> {
    let (fm, body) = frontmatter::split(content);
    let mut tags = extract_tags_from_body(body);
    if let Some(yaml_tags) = fm.tags.as_ref() {
        for t in yaml_tags {
            let canon = canonicalize(t);
            if !canon.is_empty() && !tags.contains(&canon) {
                tags.push(canon);
            }
        }
    }
    tags.sort();
    tags.dedup();
    tags
}

/// Expand a tag list with every slash-prefix so `#projects/malt` matches
/// queries for `tag:projects`. Used at index time only.
pub fn expand_with_slash_parents(tags: &[String]) -> Vec<String> {
    let mut out: Vec<String> = Vec::new();
    for t in tags {
        out.push(t.clone());
        let mut prefix = String::new();
        for part in t.split('/') {
            if !prefix.is_empty() {
                prefix.push('/');
            }
            prefix.push_str(part);
            if prefix != *t {
                out.push(prefix.clone());
            }
        }
    }
    out.sort();
    out.dedup();
    out
}

/// Normalize a tag string (strip leading #, lowercase, trim invalid chars
/// off the ends). Returns "" if the result isn't a valid tag.
pub fn canonicalize(raw: &str) -> String {
    let s = raw.trim().trim_start_matches('#').to_lowercase();
    if s.is_empty() {
        return String::new();
    }
    let chars: Vec<char> = s.chars().collect();
    if !chars[0].is_ascii_alphabetic() {
        return String::new();
    }
    let cleaned: String = chars
        .into_iter()
        .take_while(|c| c.is_ascii_alphanumeric() || matches!(c, '-' | '_' | '/'))
        .collect();
    cleaned
}

/// Strip `[[X]]` brackets, preserving inner text. The AI sees prose, not
/// markup. Same-line regex equivalent (multi-line wikilinks aren't a thing).
pub fn strip_wikilink_brackets(text: &str) -> String {
    let bytes = text.as_bytes();
    let mut out = String::with_capacity(text.len());
    let mut i = 0usize;
    while i < bytes.len() {
        if i + 1 < bytes.len() && bytes[i] == b'[' && bytes[i + 1] == b'[' {
            // Look for matching ]] on same line, no nested [.
            let mut j = i + 2;
            let mut found = false;
            while j + 1 < bytes.len() {
                if bytes[j] == b'\n' || bytes[j] == b'[' {
                    break;
                }
                if bytes[j] == b']' && bytes[j + 1] == b']' {
                    found = true;
                    break;
                }
                j += 1;
            }
            if found {
                // Emit inner [i+2..j].
                out.push_str(&text[i + 2..j]);
                i = j + 2;
                continue;
            }
        }
        // Pass through one char (handle multi-byte).
        let rest = &text[i..];
        if let Some(c) = rest.chars().next() {
            out.push(c);
            i += c.len_utf8();
        } else {
            break;
        }
    }
    out
}

/// Strip the canonical tag line + every inline `#hashtag` + every `[[...]]`
/// bracket from `text`. Used by the AI call paths so our private markup
/// doesn't pollute the LLM context. Mirrors the TS `stripTagsForAI`.
pub fn strip_tags_for_ai(text: &str) -> String {
    // Strip wikilink brackets first so subsequent passes operate on prose.
    let text = strip_wikilink_brackets(text);
    let mut lines: Vec<String> = text.split('\n').map(String::from).collect();

    // Drop the canonical tag line if it ends `text`.
    let mut last_nonempty: Option<usize> = None;
    for (i, l) in lines.iter().enumerate().rev() {
        if !l.trim().is_empty() {
            last_nonempty = Some(i);
            break;
        }
    }
    if let Some(idx) = last_nonempty {
        if is_canonical_tag_line(&lines[idx]) {
            lines.remove(idx);
            if idx > 0 && lines[idx - 1].trim().is_empty() {
                lines.remove(idx - 1);
            }
        }
    }
    let joined = lines.join("\n");

    // Remove every inline hashtag, collapsing adjacent whitespace.
    let inline = inline_tag_positions(&joined);
    let mut out = String::with_capacity(joined.len());
    let mut cursor = 0usize;
    for (from, to) in inline {
        out.push_str(&joined[cursor..from]);
        let after = joined.as_bytes().get(to).copied();
        if matches!(after, Some(b' ') | Some(b'\t')) {
            cursor = to + 1;
        } else if out.ends_with(' ') || out.ends_with('\t') {
            out.pop();
            cursor = to;
        } else {
            cursor = to;
        }
    }
    out.push_str(&joined[cursor..]);

    // Trim trailing whitespace per line, collapse blank-line runs.
    let mut result = out
        .split('\n')
        .map(|l| l.trim_end().to_string())
        .collect::<Vec<_>>()
        .join("\n");
    while result.contains("\n\n\n") {
        result = result.replace("\n\n\n", "\n\n");
    }
    result
}

/// Return (from, to) byte offsets of every inline #tag in `text`. Mirrors
/// scan_line but returns positions, not strings.
pub fn inline_tag_positions(text: &str) -> Vec<(usize, usize)> {
    let mut out = Vec::new();
    let mut in_fence = false;
    let mut line_start = 0usize;
    let bytes = text.as_bytes();
    let mut i = 0usize;
    while i <= bytes.len() {
        if i == bytes.len() || bytes[i] == b'\n' {
            let line = &text[line_start..i];
            let trimmed = line.trim_start();
            if trimmed.starts_with("```") || trimmed.starts_with("~~~") {
                in_fence = !in_fence;
            } else if !in_fence {
                scan_line_positions(line, line_start, &mut out);
            }
            line_start = i + 1;
        }
        i += 1;
    }
    out
}

fn scan_line_positions(line: &str, line_offset: usize, out: &mut Vec<(usize, usize)>) {
    let masked = strip_inline_code(line);
    let bytes = masked.as_bytes();
    let mut i = 0usize;
    while i < bytes.len() {
        if bytes[i] == b'#' {
            let prev = if i == 0 { None } else { Some(bytes[i - 1]) };
            let valid_lead = match prev {
                None => true,
                Some(b) => (b as char).is_whitespace() || matches!(b, b'(' | b'[' | b'{' | b'"' | b'\''),
            };
            if valid_lead {
                let start = i + 1;
                if start < bytes.len() && (bytes[start] as char).is_ascii_alphabetic() {
                    let mut j = start + 1;
                    while j < bytes.len()
                        && ((bytes[j] as char).is_ascii_alphanumeric()
                            || matches!(bytes[j], b'-' | b'_' | b'/'))
                    {
                        j += 1;
                    }
                    let mut end = j;
                    while end > start && bytes[end - 1] == b'/' {
                        end -= 1;
                    }
                    if end > start {
                        out.push((line_offset + i, line_offset + end));
                    }
                    i = j;
                    continue;
                }
            }
        }
        i += 1;
    }
}

/// True if `line` is *only* hashtags (and whitespace). Used to detect the
/// canonical tag line at the bottom of the body. Currently called from
/// tests + frontend (via IPC); silence dead-code while we wire things up.
#[allow(dead_code)]
pub fn is_canonical_tag_line(line: &str) -> bool {
    let trimmed = line.trim();
    if trimmed.is_empty() {
        return false;
    }
    let mut saw_tag = false;
    for tok in trimmed.split_whitespace() {
        if !tok.starts_with('#') {
            return false;
        }
        let body = &tok[1..];
        if body.is_empty() {
            return false;
        }
        let mut chars = body.chars();
        if !chars.next().map(|c| c.is_ascii_alphabetic()).unwrap_or(false) {
            return false;
        }
        if !chars.all(|c| c.is_ascii_alphanumeric() || matches!(c, '-' | '_' | '/')) {
            return false;
        }
        saw_tag = true;
    }
    saw_tag
}

// ---------- internal helpers ----------

fn strip_inline_code(line: &str) -> String {
    // Replace runs between matched backticks with spaces so positions don't
    // shift relative to the original line. We only care about exclusion, not
    // re-rendering, so this is fine.
    let mut out = String::with_capacity(line.len());
    let mut in_code = false;
    for c in line.chars() {
        if c == '`' {
            in_code = !in_code;
            out.push(' ');
        } else if in_code {
            out.push(' ');
        } else {
            out.push(c);
        }
    }
    out
}

fn scan_line(line: &str) -> Vec<String> {
    let chars: Vec<char> = line.chars().collect();
    let mut out = Vec::new();
    let mut i = 0;
    while i < chars.len() {
        if chars[i] == '#' && is_valid_lead(&chars, i) {
            let start = i + 1;
            if start < chars.len() && chars[start].is_ascii_alphabetic() {
                let mut j = start + 1;
                while j < chars.len()
                    && (chars[j].is_ascii_alphanumeric() || matches!(chars[j], '-' | '_' | '/'))
                {
                    j += 1;
                }
                // Strip a trailing `/` (e.g. `#a/`) — partial nesting, not a tag.
                let mut end = j;
                while end > start && chars[end - 1] == '/' {
                    end -= 1;
                }
                if end > start {
                    let tag: String = chars[start..end].iter().collect::<String>().to_lowercase();
                    out.push(tag);
                }
                i = j;
                continue;
            }
        }
        i += 1;
    }
    out
}

fn is_valid_lead(chars: &[char], i: usize) -> bool {
    if i == 0 {
        return true;
    }
    let prev = chars[i - 1];
    if prev.is_whitespace() {
        return true;
    }
    matches!(prev, '(' | '[' | '{' | '"' | '\'')
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extracts_simple_inline() {
        let tags = extract_tags_from_body("Hello #world today");
        assert_eq!(tags, vec!["world"]);
    }

    #[test]
    fn ignores_digit_start() {
        let tags = extract_tags_from_body("Issue #1 reported");
        assert!(tags.is_empty());
    }

    #[test]
    fn ignores_markdown_heading() {
        let tags = extract_tags_from_body("# heading\nbody");
        assert!(tags.is_empty());
    }

    #[test]
    fn handles_nesting() {
        let tags = extract_tags_from_body("#projects/malt is going well");
        assert_eq!(tags, vec!["projects/malt"]);
    }

    #[test]
    fn skips_code_fence() {
        let body = "before #foo\n```\n#bar in code\n```\n#baz after";
        let tags = extract_tags_from_body(body);
        assert_eq!(tags, vec!["baz", "foo"]);
    }

    #[test]
    fn skips_inline_code() {
        let tags = extract_tags_from_body("Use `#tag` like this #yes");
        assert_eq!(tags, vec!["yes"]);
    }

    #[test]
    fn skips_url_fragments() {
        let tags = extract_tags_from_body("See https://example.com/page#section for details");
        assert!(tags.is_empty());
    }

    #[test]
    fn canonical_tag_line() {
        assert!(is_canonical_tag_line("#foo #bar"));
        assert!(is_canonical_tag_line("  #foo/bar  "));
        assert!(!is_canonical_tag_line("#foo and bar"));
        assert!(!is_canonical_tag_line("just text"));
        assert!(!is_canonical_tag_line(""));
    }

    #[test]
    fn slash_parents() {
        let expanded = expand_with_slash_parents(&["projects/malt/v0".to_string()]);
        assert_eq!(
            expanded,
            vec![
                "projects".to_string(),
                "projects/malt".to_string(),
                "projects/malt/v0".to_string(),
            ]
        );
    }
}
